### DISCUSSION TITLE
Event after the state calls the invoke function

### DISCUSSION THREAD

**Discussion by oom-** [statelyai/xstate#3262] [Q&A] [answered]
Hello, 

I'm using XState to write Bots
They are working really well, but we are wishing to display extra information after they perform specific actions.

**For example:** 
![image](https://user-images.githubusercontent.com/11071671/166120544-23375bd8-842e-401f-b951-e594c49494db.png)

When the robot goes to the state "getSample", it invokes a promise function that is fetching some data, maps them to a JSON object, and adds it to the context.

I know when the Bot is "transitioning" because of the "onTransition" event. But this event is fired when the "state" just changed, not when the promise end, meaning that the context I get inside the "onTransition" event is not up to date.

Is there any event on which I can subscribe that will be fired at the conclusion of the invoke function ? something like "onTransitionDone" or something?

Thank you a lot.

**Comment by davidkpiano** [comment]
> Is there any event on which I can subscribe that will be fired at the conclusion of the invoke function ? something like "onTransitionDone" or something?

Yes, you can use `waitFor` to wait for some condition to be true:

```js
import { waitFor } from 'xstate/lib/waitFor';

// ...
const initializedState = await waitFor(service, state => state.matches('initialized'));
```

I recommend adding an `onDone` transition to the `invoke` config that handles what happens when the invocation is done.

**Comment by Andarist** [comment, marked as answer]
Another solution would be to add your callback to `invoke.onDone`. That would make your logic visible on the diagram itself - whereas when you listen for stuff like this "outside" of a machine you are bypassing the modeled behavior.

**Reply by oom-** [reply]
Sorry for my late answer. I'm gonna try it like that. I want to keep it as much "inside" of the machine as possible.
Thank you.
